### PR TITLE
feat: [alpha] basic hcaptcha support

### PR DESCRIPTION
**Pull request by darora**
## What kind of change does this PR introduce?

Adds support for HCaptcha protection on routes that send out emails/SMS. Should be considered alpha - API might be changed in the future.

## What is the current behavior?

n/a

## What is the new behavior?

If HCaptcha support is enabled, and one is not provided as part of the request, the request is rejected.

### CODE REVIEW

**Comment by darora** [comment]
Verified via manual testing (axios); opened a accompanying PR on gotrue-js but that wasn't used to verify this change

File: security/hcaptcha.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type CaptchaResponse struct {
+	Token string `json:"hcaptchaToken"`
```
Comment (kangmingtay) [inline security/hcaptcha.go:18]: nit: we use snake_case instead of camelCase in gotrue for json fields

File: security/hcaptcha.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+type VerificationResponse struct {
+	Success    bool     `json:"success"`
+	ErrorCodes []string `json:"error-codes"`
+	Hostname   string   `json:"hostname"`
+}
+
```
Comment (kangmingtay) [inline security/hcaptcha.go:26]: same as above

File: security/hcaptcha.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	"time"
+)
+
+type CaptchaResponse struct {
+	Token string `json:"hcaptchaToken"`
+}
+
+type VerificationResponse struct {
+	Success    bool     `json:"success"`
+	ErrorCodes []string `json:"error-codes"`
```
Comment (kangmingtay) [inline security/hcaptcha.go:23]: same as above

File: security/hcaptcha.go (hunk 1 above)
Comment (kangmingtay) [reply security/hcaptcha.go:18]: should we also enforce users to pass the site_key as well? seems like it's recommended on the hcaptcha docs (https://docs.hcaptcha.com/#verify-the-user-response-server-side)

Comment (kangmingtay) [reply security/hcaptcha.go:18]: we can also get the client's ip address using `r.RemoteAddr` and pass it in the POST request to hcaptcha too

**Comment by kangmingtay** [comment]
Probably not the highest priority right now, but to extend the support for other captcha providers, we could declare an interface as such:
```
type CaptchaProvider interface {
  VerifyRequest(r *http.Request, secretKey string) (VerificationResult, error) 
}
```

and do something like [this](https://github.com/supabase/gotrue/blob/9b955cbce0f19979d32429a3e5bb5425e7fd05f2/api/external.go#L334) to get the captcha provider

File: security/hcaptcha.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	Hostname   string   `json:"hostname"`
+}
+
+type VerificationResult int
+const (
+	UserRequestFailed VerificationResult = iota
```
Comment (kangmingtay) [inline security/hcaptcha.go:29]: oh wow i never knew you could do this to initialise enums haha 👀

**Comment by kangmingtay** [comment]
this is gonna be such a useful feature once it's merged! 🔥

File: security/hcaptcha.go (hunk 1 above)
Comment (darora) [reply security/hcaptcha.go]: Renamed to  use snake case, and started passing client IP along. I've intentionally skipped site key for now; it's not necessary to get value out of this chagne, and I'm not sure what we want support for that to look like (e.g. keyed on the origin to support multiple keys? some other metric?)

Comment (darora) [reply security/hcaptcha.go]: (skipping abstracting out things to support other providers for the same reason)

**Review by inian — APPROVED** [review]

File: security/hcaptcha.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	Client = &http.Client{Timeout: 10 * time.Second}
+}
+
+func VerifyRequest(r *http.Request, secretKey string) (VerificationResult, error) {
+	res := CaptchaResponse{}
+	bodyBytes, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return UserRequestFailed, err
+	}
+	r.Body.Close()
```
Comment (kangmingtay) [inline security/hcaptcha.go:46]: nit: defer this?

File: security/hcaptcha.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+		return VerificationProcessFailure, errors.Wrap(err, "couldn't initialize request object for hcaptcha check")
+	}
+	r.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	r.Header.Add("Content-Length", strconv.Itoa(len(data.Encode())))
+	res, err := Client.Do(r)
+	if err != nil {
+		return VerificationProcessFailure, errors.Wrap(err, "failed to verify hcaptcha token")
+	}
+	verResult := VerificationResponse{}
+	decoder := json.NewDecoder(res.Body)
```
Comment (kangmingtay) [inline security/hcaptcha.go:77]: necessary to `defer res.Body.Close()` here? seems like it's recommended in the go package [here](https://pkg.go.dev/net/http)

**Review by kangmingtay — APPROVED** [review]

File: security/hcaptcha.go (hunk 6 above)
Comment (darora) [reply security/hcaptcha.go:78]: done

File: security/hcaptcha.go (hunk 5 above)
Comment (darora) [reply security/hcaptcha.go:46]: There's no reason to defer it, it's replaced in the next statement, and has already been consumed

File: api/middleware_test.go [hunk 7]
```diff
@@ -1,12 +1,85 @@
...
 
+	"github.com/gofrs/uuid"
+	"github.com/netlify/gotrue/conf"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
 )
 
+const (
+	secret   string = "0x0000000000000000000000000000000000000000"
```
Comment (darora) [inline api/middleware_test.go:19]: suggest prefixing var names with `hcaptcha_` to make things clear

**Pull request merged**